server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -2919,45 +2920,65 @@ private void doSnapshotShard(SnapshotShardContext context) {
-                        blobStoreIndexShardSnapshot,
-                        shardContainer,
-                        snapshotUUID,
-                        compress,
-                        serializationParams
+                    try {
+                        final String snapshotUUID = snapshotId.getUUID();
+                        final Map<String, String> serializationParams = Collections.singletonMap(
+                            BlobStoreIndexShardSnapshot.FileInfo.SERIALIZE_WRITER_UUID,
+                            Boolean.toString(writeFileInfoWriterUUID)
+                        );
+                        INDEX_SHARD_SNAPSHOT_FORMAT.write(
+                            blobStoreIndexShardSnapshot,
+                            shardContainer,
+                            snapshotUUID,
+                            compress,
+                            serializationParams
+                        );
+                    } catch (IOException e) {
+                        throw new IndexShardSnapshotFailedException(shardId, "Failed to write commit point", e);
+                    }
+                    afterWriteSnapBlob.run();
+                    final ShardSnapshotResult shardSnapshotResult = new ShardSnapshotResult(
+                        indexGeneration,
+                        ByteSizeValue.ofBytes(blobStoreIndexShardSnapshot.totalSize()),
+                        getSegmentInfoFileCount(blobStoreIndexShardSnapshot.indexFiles())
                     );
-                } catch (IOException e) {
-                    throw new IndexShardSnapshotFailedException(shardId, "Failed to write commit point", e);
+                    snapshotStatus.moveToDone(threadPool.absoluteTimeInMillis(), shardSnapshotResult);
+                    context.onResponse(shardSnapshotResult);
+                } catch (Exception e) {
+                    // If failure occurs here, no cleanup will be performed because it might be possible that
+                    // written files are referenced by another concurrent process.
+                    context.onFailure(e);
                 }
-                afterWriteSnapBlob.run();
-                final ShardSnapshotResult shardSnapshotResult = new ShardSnapshotResult(
-                    indexGeneration,
-                    ByteSizeValue.ofBytes(blobStoreIndexShardSnapshot.totalSize()),
-                    getSegmentInfoFileCount(blobStoreIndexShardSnapshot.indexFiles())
-                );
-                snapshotStatus.moveToDone(threadPool.absoluteTimeInMillis(), shardSnapshotResult);
-                delegate.onResponse(shardSnapshotResult);
+            }, e -> {
+                try {
+                    shardContainer.deleteBlobsIgnoringIfNotExists(
+                        copyOfFilesToSnapshot.stream()
+                            .flatMap(fileInfo -> IntStream.range(0, fileInfo.numberOfParts()).mapToObj(fileInfo::partName))
+                            .iterator()
```

Comment:
Nit: streams are kinda heavyweight, we can avoid them here:

```suggestion
                        Iterators.flatMap(copyOfFilesToSnapshot.iterator(), f -> Iterators.forRange(0, f.numberOfParts(), f::partName))
```

Reply:
TIL